### PR TITLE
Fix ingress usage

### PR DESCRIPTION
**Pull request by sed-i**
(Note: this PR was split out of #340.)

## Issue
The self-scrape job is "down" when prometheus is related to traefik.The external hostname is not passed to prom_scrape constructor for the self-scraping endpoint.

## Solution
Use external_url in constructing the self-scraping job.

Note: the quality of the solution in this PR is affected by https://github.com/canonical/traefik-k8s-operator/issues/78.

## Testing Instructions
Deploy prom and relate to traefik. Then look at prom's `/api/v1/targets`.

## Release Notes
Fix ingress usage.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -308,6 +311,11 @@ def _update_status(self, event):
         """Fired intermittently by the Juju agent."""
         self.unit.set_workload_version(self._prometheus_server.version())
 
+        # Call `_configure`` to resolve "BlockedStatus" when the ready-for-unit event is processed
+        # before ingress is in fact ready.
+        # https://github.com/canonical/traefik-k8s-operator/issues/78
+        self._configure(event)
```
Comment (sed-i) [inline src/charm.py:317]: Copying comment by @rbarry82 from https://github.com/canonical/prometheus-k8s-operator/pull/340#pullrequestreview-1093589121:

> -   Update `IngressPerUnitRequirer` logic. This seems to be most of the changes in `src/charm.py`  
>     - Even there, I would suggest that "call `_configure|_common_exit_hook` hoping it will resolve `BlockedStatus` if an event was emitted before ingress is ready is a sledgehammer which is bad considering we already have an issue about `/-/reload` being called to often, what whatever logic it would take for "is `BlockedStatus` set during `_update_status`? Is it still valid? If it is, clear it, if not, move on) would be more explicit, cleaner, and more legible.
>     - Even seeing the comment and reading `_configure`, I have no idea why this would be true. Either `current_services == new_layer.services` and it's never restarted at all, so we never try to connect, or the change to `set_workload_version` by calling the binary rather than an endpoint means that the presumptive bypass by calling `self._prometheus_server.version()` will no longer call the endpoint, and this won't actually be set.
>     - In reading through this again, it also seems wrong that `# Reload may fail if ingress ...` wold lead to a `BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE)`. It's not corrupt, though. Ingress is not ready, or Prometheus has taken too long to come up. This could potentially lead administrators down a "blind alley"
>     - The "right" fix for this is likely:
>         - Add a `can_connect()` method to `prometheus_server.Prometheus()`. If `update_status` fires and it's `BlockedStatus` with a message (a _new_ message which indicates the actual state) about timeouts:
> 
> ```
> def _update_status(self, event):
>   ...
>   if isinstance(self.unit.status, BlockedStatus) and self.unit.status.message == PROMETHEUS_TIMEOUT:
>     if self._prometheus_server.can_connect():
>       self.unit.status = ActiveStatus()
> 
> ```
> - In addition, `prometheus_server.Prometheus.*` should provide distinctions between timeouts and `ConnectionError` to flag the "real" status. Custom exceptions instead of a boolean are a trivial, Pythonic way to do this

Comment (sed-i) [reply src/charm.py:317]: > "call `_configure|_common_exit_hook` hoping it will resolve BlockedStatus if an event was emitted before ingress is ready is a sledgehammer which is bad considering we already have an issue about `/-/reload` being called to often

Agreed, but to limit the scope of this change, and given #342 will be addressed very soon, can live with this for now?

> what whatever logic it would take for "is BlockedStatus set during _update_status? Is it still valid? If it is, clear it, if not, move on) would be more explicit, cleaner, and more legible.

This would be in direct contrast to the common exit hook pattern. I'd rather not mix both approaches in the same charm. Also, once #342 is fixed this will be much less painful to look at.

> seems wrong that # Reload may fail if ingress ... wold lead to a `BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE)`

Agreed. I don't particularly like that status comparison pattern and it should be fixed as part of #342.

> prometheus_server.Prometheus.* should provide distinctions between timeouts and ConnectionError to flag the "real" status.

Sounds presumptuous at first glance. And what will the charm do with the various exceptions? Just choose one message over another? Maybe, if we have insight, expose a `retry` setting the charm could use?

Comment (sed-i) [reply src/charm.py:317]: I mentioned this on #342.

Comment (rbarry82) [reply src/charm.py:317]: > > This would be in direct contrast to the common exit hook pattern. I'd rather not mix both approaches in the same charm. Also, once #342 is fixed this will be much less painful to look at.

I think that's a little bit the point -- the common exit hook cannot cleanly address this, but we actually do need to do so. There are multiple non-`_configure` places in the charm where we set `BlockedStatus` (in `_generate_command`, `_on_k8s_...`). "Un-setting" it isn't a massive divergence.

The odds of this being needed inside this patch are low, since ingress shouldn't fail/be unavailable long enough for us to want to avoid blocking a charm in `executing` for some timeout (2-5 seconds or whatever), though.
 
> Agreed. I don't particularly like that status comparison pattern and it should be fixed as part of #342.

Why would it be changed there? It also affects this. There seems to be three cases where the charm may not be accessible:

* Bad configuration (we can check with `promtool` if there's an exception raised trying to connect)
* Timeout because it's replaying/`mmap`ing HEAD chunks which haven't been written to the TSDB
* Not reachable due to ingress

`/-/reload` being called to often doesn't have anything to do with the last case, but this patch does.

> Sounds presumptuous at first glance. And what will the charm do with the various exceptions? Just choose one message over another? Maybe, if we have insight, expose a `retry` setting the charm could use?

Using exceptions for flow control is Pythonic, and a narrow custom exception which is caught and sets an appropriate message is much less presumptuous/magic than returning a boolean from `prometheus_server.Prometheus` which may set a meaningless message which will hopefully automagically clear itself the next time something fires and hits the common exit hook.

The exceptions could have a property with a message, or a lookup table in the charm, or whatever. The charm would set a status message to tell the administrator what is happening and why Prometheus may not be reachable which is more useful than a "configuration is invalid" message which may not be true or an `ActiveStatus` when Prometheus is actually not active/reachable.

A `retry` mechanism can only be so useful. For ingress, if we retry for 2-3 seconds, it may be enough. Traefik shouldn't take that long to restart. For `/-/reload`, though, it won't be, and how we address that can certainly be discussed elsewhere (we already "know" that we'll probably have to sidestep it with @manadart's work to trigger async hooks), but we're well enough aware of the "basic" failure cases, and we need a way to flag the administrator about what the actual status is. Since this patch is quite small/contained, and is the logical place to indicate what "goes wrong" when Ingress isn't available yet (timeout or default traefik landing page, I'm not sure which), the initial scaffolding of "raise an exception of the appropriate type" seems that it should go here and not be deferred.

File: tests/integration/test_external_url.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        # )
+        ops_test.model.add_relation(f"{prometheus_app_name}:ingress", "traefik"),
+        ops_test.model.wait_for_idle(
+            apps=app_names,
+            status="active",
+            wait_for_units=num_units,
+            timeout=600,
+            idle_period=idle_period,
+        ),
+    )
```
Comment (dstathis) [inline tests/integration/test_external_url.py:84]: Shouldn't these be run sequentially rather than in parallel? When run in parallel, `wait_for_idle` could in theory return before `add_relation` starts running.

File: tests/integration/test_external_url.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    logger.info("Attmpting to fetch targets from these urls: %s", prom_urls)
+    targets_per_unit = [
+        urllib.request.urlopen(url, None).read().decode("utf8") for url in prom_urls
+    ]
+    assert all('"health":"down"' not in response for response in targets_per_unit)
+
+
+@pytest.mark.abort_on_fail
+async def test_self_scrape_job_is_up_with_config_option_overriding_traefik(ops_test: OpsTest):
+    pass
```
Comment (dstathis) [inline tests/integration/test_external_url.py:99]: I think you forgot to remove this placeholder.

**Review by dstathis — APPROVED** [review]

File: tests/integration/test_external_url.py (hunk 3 above)
Comment (sed-i) [reply tests/integration/test_external_url.py]: Done.

File: tests/integration/test_external_url.py (hunk 2 above)
Comment (sed-i) [reply tests/integration/test_external_url.py]: Done.

File: src/charm.py [hunk 4]
```diff
@@ -159,11 +171,14 @@ def log_level(self):
     @property
     def _default_config(self):
         """Default configuration for the Prometheus workload."""
+        base_path = urlparse(self.external_url).path
+        metrics_path = base_path.rstrip("/") + "/metrics"
```
Comment (rbarry82) [inline src/charm.py:175]: This is almost identical to what's in `self_scraping_job` but structured slightly differently. If there a reason not to just use one code variant and make it a property?

File: tox.ini [hunk 5]
```diff
@@ -71,9 +71,13 @@ deps =
     lib: ops
     charm: responses==0.20.0
     charm: httpcore==0.14.7
+    unit: {[testenv:unit]deps}
+    integration: {[testenv:integration]deps}
 commands =
     charm: mypy {[vars]src_path} {posargs}
     lib: mypy --python-version 3.5 {[vars]lib_path} {posargs}
+    unit: mypy {[vars]tst_path}/unit {posargs}
+    integration: mypy {[vars]tst_path}/integration {posargs}
```
Comment (rbarry82) [inline tox.ini:80]: Why are these added back?

Comment (sed-i) [reply tox.ini:80]: I think those were removed by mistake. The correct change is
```diff
-envlist = lint, static-{charm,lib,unit,integration}, unit
+envlist = lint, static-{charm,lib}, unit
```
I've done it in this PR here.

As for the envs themselves - I use them! :)

File: src/charm.py (hunk 4 above)
Comment (sed-i) [reply src/charm.py:186]: Good catch! Fixed in next commit.

File: src/charm.py [hunk 6]
```diff
@@ -88,25 +88,31 @@ def __init__(self, *args):
...
+        external_url = urlparse(self.external_url)
+
+        # TODO when https://github.com/canonical/traefik-k8s-operator/issues/78 is fixed, change
+        #  the refresh event to self.ingress.on.ready_for_unit.
         self._scraping = MetricsEndpointProvider(
             self,
             relation_name="self-metrics-endpoint",
             jobs=self.self_scraping_job,
+            external_hostname=external_url.netloc,
+            refresh_event=self.on.update_status,  # needed for ingress
```
Comment (Abuelodelanada) [inline src/charm.py:104]: Shouldn't we set more refresh_events, like??

```python
            refresh_event=[
                self.ingress_per_unit.on.ready_for_unit,
                self.ingress_per_unit.on.revoked_for_unit,
                self.on.ingress_relation_departed,
                self.on.update_status,
            ],
```

**Review by Abuelodelanada — APPROVED** [review]

File: src/charm.py (hunk 6 above)
Comment (sed-i) [reply src/charm.py]: Fixed, thanks!

**Review by Abuelodelanada — APPROVED** [review]

File: tests/unit/test_endpoint_consumer.py [hunk 7]
```diff
@@ -586,3 +590,150 @@ def wildcard_targets(jobs, wildcard_ports):
...
+            "remote-app/1",
+            {
+                "prometheus_scrape_unit_address": "11.11.11.11",
+                "prometheus_scrape_unit_path": external_url_path.substitute(unit=1),
+                "prometheus_scrape_unit_name": "remote-app/1",
+            },
+        )
+
+    def setUp(self):
+        metadata_file = open("metadata.yaml")
```
Comment (beliaev-maksim) [inline tests/unit/test_endpoint_consumer.py:644]: this is dependent on the invocation directory
